File: basics/02-linkedlist/0206-142.cpp
/*
 * leetcode 142
 * detect cycle: linked list cycle II
 *
 */

struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};

class Solution {
public:
    ListNode *detectCycle(ListNode *head) {
        ListNode* fast = head;
        ListNode* slow = head;
        while (fast != nullptr && fast->next != nullptr) {
            fast = fast->next->next; // two steps
            slow = slow->next; // one step
            if (fast == slow) break;
        }
        if (fast == nullptr || fast->next == nullptr) {
            // no cycle
            return nullptr;
        }
        // restart for slow
        slow = head;
        while(slow != fast) {
            fast = fast->next;
            slow = slow->next;
        }
        return slow;
    }
};

File: basics/02-linkedlist/0206-142.go
/*
 * leetcode 142
 */

type ListNode struct {
    Val int
    Next *ListNode
}

func detectCycle(head *ListNode) *ListNode {
    var slow *ListNode
    var fast *ListNode
    if head != nil && head.Next != nil {
        slow = head.Next
        fast = head.Next.Next
    } else {
        return nil
    }
    for fast != nil && fast.Next != nil {
        slow = slow.Next
        fast = fast.Next.Next
        if slow == fast {
            for head!= fast {
                head = head.Next
                fast = fast.Next
            }
            return head
        }
    }
    return nil
}
